### PR TITLE
feat: add an ONNX adapter for WASM

### PR DESCRIPTION
**Pull request by dmadisetti**
## Summary
Adds a ONNX runtime adapter for `jax` and `pytorch` models such that they can run inference over WASM.

Example:

```python
from moutils.onnx import OnnxRuntime

runtime = OnnxRuntime.from_torch(model, (example_input,), output_names=["y"])
runtime = OnnxRuntime.from_jax(fn, [("B", 4)], input_names=["x"])  # via jax2onnx
runtime = OnnxRuntime(onnx_bytes)                                  # low-level

logits = (await runtime.run({"x": x}, ["y"]))[0]   # async; picks the backend
```
## Also included

- `notebooks/onnx_mnist1d.py` end-to-end example (MLP trained on MNIST-1D, cached as an OnnxRuntime, lasso-driven live inference), verified to export and restore torch-free in a WASM build.
- `README` update for an "ONNX runtime adapter" section documenting the required pattern.

## Installation

```shell
pip install "moutils[wasm]"
```

The module lazily imports all external dependencies, and has some basics handling for WASM vs native runtimes.

### CODE REVIEW

File: tests/test_onnx.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    )
+    model = helper.make_model(graph, opset_imports=[helper.make_opsetid("", 13)])
+    return model.SerializeToString()
+
+
+def test_native_run_executes_model():
+    pytest.importorskip("onnxruntime")
+    rt = OnnxRuntime(_double_model_bytes())
+    x = np.array([[1.0, 2.0], [3.0, 4.0]], dtype=np.float32)
+    (y,) = asyncio.run(rt.run({"x": x}, ["y"]))
```
[the remark this answers is not in the document]
Comment (dmadisetti) [reply tests/test_onnx.py:65]: I think this would be a lot of boiler plate

File: src/moutils/onnx.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    ```python
+    from moutils.onnx import OnnxRuntime
+
+    # From a torch model (torch.onnx.export kwargs pass through):
+    runtime = OnnxRuntime.from_torch(model, (example_input,), output_names=["y"])
+    # Or wrap serialized ONNX bytes directly:
+    runtime = OnnxRuntime(onnx_bytes)
+
+    logits = (await runtime.run({"x": x}))[0]  # x: a numpy input array
+    ```
```
Comment (akshayka) [inline src/moutils/onnx.py]: Nice docs, but move them to the class docstring so it shows up in our completions tooltips

File: src/moutils/onnx.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+Importing this module registers a marimo cache stub, so a cached `OnnxRuntime`
+round-trips as exactly its model bytes. A torch or jax model trained at export
+time then restores as a working inference session inside a static WASM export.
+
+Example:
+    ```python
+    from moutils.onnx import OnnxRuntime
+
+    # From a torch model (torch.onnx.export kwargs pass through):
+    runtime = OnnxRuntime.from_torch(model, (example_input,), output_names=["y"])
```
Comment (akshayka) [inline src/moutils/onnx.py]: What is `example_input`? I can guess but it's good to be explicit

File: src/moutils/onnx.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        cls,
+        model: Any,
+        args: Any,
+        *,
+        wasm_version: str = DEFAULT_WASM_VERSION,
+        **export_kwargs: Any,
+    ) -> OnnxRuntime:
+        """Export a torch model to ONNX and wrap the bytes.
+
+        `args` and `export_kwargs` pass through to `torch.onnx.export`.
```
Comment (akshayka) [inline src/moutils/onnx.py]: I wouldn't assume people are familiar with ONNX. Explain what the arguments are. Also `export_kwargs` doesn't exist on on the method?

File: src/moutils/onnx.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        inputs: Any,
+        *,
+        wasm_version: str = DEFAULT_WASM_VERSION,
+        **to_onnx_kwargs: Any,
+    ) -> OnnxRuntime:
+        """Convert a JAX function to ONNX and wrap the bytes.
+
+        Uses [`jax2onnx`](https://pypi.org/project/jax2onnx/). `inputs` (a
+        sequence of shapes) and `to_onnx_kwargs` pass through to
+        `jax2onnx.to_onnx`.
```
Comment (akshayka) [inline src/moutils/onnx.py]: Include a usage example?

File: src/moutils/onnx.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+            self._session = ort.InferenceSession(self.onnx_bytes)
+            self._kind = "native"
+
+    async def run(
+        self,
+        inputs: dict[str, Any],
+        output_names: list[str] | None = None,
+    ) -> list[Any]:
+        """Run inference and return the outputs in `output_names` order.
```
Comment (akshayka) [inline src/moutils/onnx.py]: Usage example here too? Not totally clear how to use `output_names`

File: README.md [hunk 7]
```diff
@@ -446,6 +446,75 @@ These connections do not support bound parameters. Use static or trusted SQL.
...
+## ONNX runtime adapter
+
+`moutils.onnx.OnnxRuntime` wraps a serialized ONNX model in a lazy inference
+session and registers a marimo cache stub for it. When marimo caches an
+`OnnxRuntime`, it stores exactly the model bytes, not the training framework's
+objects.
+
+This lets a WASM export ship a trained model and run inference in the browser
+without the training framework. Enable cell caching in the notebook header. Then
+build the runtime in a cell:
```
Comment (akshayka) [inline README.md]: Flip these paragraphs? Basically explain what this runtime adapter is first and why it's useful, then describe the implementation

**Review by akshayka — COMMENTED** [review]
Comments about comments  🙂

File: README.md (hunk 7 above)
Comment (dmadisetti) [reply README.md]: Agreed, tried to lead with this

File: README.md [hunk 8]
```diff
@@ -446,6 +446,74 @@ These connections do not support bound parameters. Use static or trusted SQL.
 Do not insert untrusted values into SQL strings. Provider adapters that receive
 an existing client leave that client open by default.
 
+## ONNX runtime adapter
+
+`moutils.onnx.OnnxRuntime` exposes a inference runtime for pytorch and jax
```
[the remark this answers is not in the document]
Comment (dmadisetti) [reply README.md]: Provide a suggestion?

File: README.md [hunk 9]
```diff
@@ -446,6 +446,74 @@ These connections do not support bound parameters. Use static or trusted SQL.
...
+models in WASM marimo notebooks through a
+[cacheable](https://docs.marimo.io/api/caching/) interface and the
+[ONNX runtime](https://onnxruntime.ai/). As jax and torch are not available in
+the browser, the adapter
+exports the model to ONNX and runs inference with `onnxruntime-web` when used in
+browser. See [`notebooks/onnx_mnist1d.py`](notebooks/onnx_mnist1d.py) for a
+full example- which trains a MNIST-1D model, then bundles a static page, and
+provides interactive inference.
```
Comment (dmadisetti) [inline README.md]:
```suggestion
`moutils.onnx.OnnxRuntime` exposes an inference runtime for PyTorch and JAX
models in WASM marimo notebooks through a
[cacheable](https://docs.marimo.io/api/caching/) interface and the
[ONNX runtime](https://onnxruntime.ai/). Since PyTorch and JAX aren't available
in the browser, the adapter exports the model to ONNX and runs inference with
`onnxruntime-web` when used in the browser. See
[`notebooks/onnx_mnist1d.py`](notebooks/onnx_mnist1d.py) for a full example,
which trains an MNIST-1D model, bundles a static page, and provides interactive
inference.
```

**Review by akshayka — APPROVED** [review]
